File: storage/raw_store.py
"""原始快照存储：SHA-256 内容哈希寻址，防篡改、同内容天然去重。"""
from __future__ import annotations

import hashlib
import os
from pathlib import Path


class RawStore:
    """快照按 {sha256}.md 落盘：同内容幂等（文件已存在即跳过写入）。"""

    def __init__(self, root: str | Path = "data/raw"):
        self.root = Path(root)
        self.root.mkdir(parents=True, exist_ok=True)

    @staticmethod
    def content_hash(content: str) -> str:
        return hashlib.sha256(content.encode("utf-8")).hexdigest()

    def save(self, content: str) -> str:
        """写入快照并返回内容哈希；tmp + os.replace 原子落盘，
        进程中断不会留下同名截断文件破坏 exists() 幂等语义。"""
        digest = self.content_hash(content)
        path = self.path(digest)
        if not path.exists():
            tmp = path.with_suffix(".tmp")
            tmp.write_text(content, encoding="utf-8")
            os.replace(tmp, path)
        return digest

    def path(self, content_hash: str) -> Path:
        return self.root / f"{content_hash}.md"

    def exists(self, content_hash: str) -> bool:
        return self.path(content_hash).exists()

    def read(self, content_hash: str) -> str:
        return self.path(content_hash).read_text(encoding="utf-8")

File: src/storage/rawStore.ts
/** 原始快照存储：SHA-256 内容哈希寻址，防篡改、同内容天然去重。 */
import { createHash } from "node:crypto";
import { existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from "node:fs";
import { join } from "node:path";

export class RawStore {
  readonly root: string;

  constructor(root: string = "data/raw") {
    this.root = root;
    mkdirSync(root, { recursive: true });
  }

  static contentHash(content: string): string {
    return createHash("sha256").update(content, "utf8").digest("hex");
  }

  /** 写入快照并返回内容哈希；tmp + rename 原子落盘，同内容幂等。 */
  save(content: string): string {
    const digest = RawStore.contentHash(content);
    const path = this.path(digest);
    if (!existsSync(path)) {
      const tmp = `${path}.tmp`;
      writeFileSync(tmp, content, "utf8");
      renameSync(tmp, path);
    }
    return digest;
  }

  path(contentHash: string): string {
    return join(this.root, `${contentHash}.md`);
  }

  exists(contentHash: string): boolean {
    return existsSync(this.path(contentHash));
  }

  read(contentHash: string): string {
    return readFileSync(this.path(contentHash), "utf8");
  }
}
